File: dice_game.py
import random

class DiceGame:
    def __init__(self):
        self.current_player = 'player1'
        self.p1_score = 0
        self.p2_score = 0
        self.turn_score = 0
        self.gameover = False
        self.winner = None
        
    def switch_player(self):
        self.turn_score = 0
        if self.current_player == 'player1':
            self.current_player = 'player2'
            return "Player 2's turn"
        else:
            self.current_player = 'player1'
            return "Player 1's turn"
            
            
    def roll_dice(self):
        if self.gameover:
            return "game is over, illegal move"
        roll_value = random.randint(1,6)
        if roll_value == 1:
            self.turn_score = 0
            self.switch_player()
            return f"{self.current_player} has rolled a 1 and lost their turn score"
        
        self.turn_score += roll_value
        return f"{self.current_player} rolled {roll_value} and their total score is now {self.turn_score} "
    
    def hold_dice(self):
        if self.gameover:
            return "game is over, illegal move"
        
        if self.current_player == 'player1':
            self.p1_score += self.turn_score
            
            if self.p1_score >= 100:
                self.gameover = True
                self.winner = 'player1'
                return "player 1 wins"
        else:
            self.p2_score += self.turn_score
            
            if self.p2_score >= 100:
                self.gameover = True
                self.winner = 'player2'
                return "player 2 wins"
            
    def display_scores(self):
        return {
            "current_player": self.current_player,
            "turn_score": self.turn_score,
            "player1_score": self.p1_score,
            "player2_score": self.p2_score,
            "winner": self.winner
        }
        
        
game = DiceGame()
print(game.display_scores())
print(game.roll_dice())
print(game.hold_dice())
print(game.display_scores())
print(game.roll_dice())

print(game.hold_dice())

print(game.display_scores())

print(game.roll_dice())
print(game.hold_dice())

print(game.display_scores())

File: DiceGame.js
const readline = require("readline");

class Player{
    constructor(name, unheldScore = 0, score = 0){
        this.name = name;
        this.unheldScore = unheldScore;
        this.score = score;
    }

    resetUnheldScore(){
        this.unheldScore = 0;
    }

    roll(dice){
        const rolled = dice.roll();
        console.log(`${this.name} rolled a ${rolled}`);

        if (rolled === 1){
            this.resetUnheldScore();
            console.log(`${this.name} rolled a 1 and lost unheld score. Turn switches.`);
            return false;
        }

        this.unheldScore += rolled;
        console.log(`${this.name}'s unheld score is now ${this.unheldScore}`);
        return true;
    }

    hold(){
        this.score += this.unheldScore;
        console.log(`${this.name} holds. ${this.unheldScore} points added.`);
        this.resetUnheldScore();
        console.log(`${this.name}'s total score is now ${this.score}`);
    }

    hasWon(){
        return this.score >= 100;
    }
}

class Dice{
    roll(){
        return Math.floor(Math.random() * 6) + 1;
    }
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function ask(question){
    return new Promise((resolve) => {
        rl.question(question, (answer) => {
            resolve(answer.trim().toLowerCase());
        });
    });
}

async function playGame(){
    const player1 = new Player("Jason");
    const player2 = new Player("AnaplanDev");
    const dice = new Dice();
    let currentPlayer = player1;

    console.log("Dice Game started.");
    console.log("Commands: 'r' to roll, 'h' to hold.");

    while (true){
        console.log(`\n${currentPlayer.name}'s turn.`);
        console.log(`Current total score: ${currentPlayer.score}`);
        console.log(`Current unheld score: ${currentPlayer.unheldScore}`);

        const choice = await ask("Roll or Hold? (r/h): ");

        if (choice === "r"){
            const canContinue = currentPlayer.roll(dice);

            if (!canContinue){
                currentPlayer = currentPlayer === player1 ? player2 : player1;
            }
        } else if (choice === "h"){
            currentPlayer.hold();

            if (currentPlayer.hasWon()){
                console.log(`\n${currentPlayer.name} wins with ${currentPlayer.score} points!`);
                break;
            }

            currentPlayer = currentPlayer === player1 ? player2 : player1;
        } else {
            console.log("Invalid input. Please enter 'r' or 'h'.");
        }
    }

    rl.close();
}

playGame();

